Bug: thinking stall retry exhaustion classified as builder_unknown_failure

## Problem

When the builder thinking stall retry budget is exhausted, the shepherd logs a specific error but then classifies it as \`builder_unknown_failure\` in the systematic failure tracker. This makes it impossible to distinguish genuine unknown failures from thinking stall exhaustion in diagnostics or recovery logic.

## Observed Behavior

\`\`\`
[ERROR] builder thinking stall: extended thinking output with zero tool calls detected (retry budget exhausted)
...
Systematic failure detected: pattern=builder_unknown_failure, threshold=3
[WARN] Systematic failure escalation: issue #2853 moved to loom:blocked (pattern=builder_unknown_failure, count=3)
\`\`\`

## Expected Behavior

Thinking stall exhaustion should be classified as its own error pattern (e.g. \`builder_thinking_stall\`) so that:

1. Operators can distinguish it from genuine unknown failures in systematic failure reports
2. Recovery guidance can be tailored: "consider increasing \`LOOM_BUILDER_THINKING_STALL_TIMEOUT\`" vs generic "investigate failure pattern"
3. Automatic recovery strategies (e.g. back off and retry with longer timeout) can be applied selectively

## Proposed Fix

In \`builder.py\`, where thinking stall exhaustion is detected, pass a specific \`error_class\` value to the systematic failure recorder rather than falling through to the generic path.

\`\`\`python
# Instead of generic builder_unknown_failure:
record_failure(issue, error_class="builder_thinking_stall", ...)
\`\`\`

## References

- Observed during shepherd run for #2853 (2026-02-19)
- Thinking stall detection: \`loom-tools/src/loom_tools/shepherd/phases/base.py\`
- PR #2903 increases the builder thinking stall timeout (related mitigation)

--- a/loom-tools/src/loom_tools/shepherd/cli.py
+++ b/loom-tools/src/loom_tools/shepherd/cli.py
@@ -2338,7 +2338,7 @@ def _record_fallback_failure(ctx: ShepherdContext, exit_code: int) -> None:
         record_blocked_reason,
     )
 
-    # Classify the failure based on exit code
+    # Classify the failure based on exit code and abandonment_info flags.
     if exit_code == ShepherdExitCode.SYSTEMIC_FAILURE:
         error_class = "auth_infrastructure_failure"
         details = "Builder failed due to auth/API infrastructure issue (fallback cleanup)"
@@ -2348,6 +2348,20 @@ def _record_fallback_failure(ctx: ShepherdContext, exit_code: int) -> None:
     elif exit_code == ShepherdExitCode.RATE_LIMIT_ABORT:
         error_class = "rate_limit_abort"
         details = "CLI hit usage/plan limit — interactive prompt in headless mode (fallback cleanup)"
+    elif (
+        ctx.abandonment_info is not None
+        and ctx.abandonment_info.get("failure_data", {}).get("thinking_stall")
+    ):
+        # Thinking stall exhaustion: builder produced thinking output but zero
+        # tool calls across all retries.  Classified separately from generic
+        # unknown failures so operators can distinguish it in diagnostics and
+        # recovery guidance can be tailored (e.g. "increase
+        # LOOM_BUILDER_THINKING_STALL_TIMEOUT").  See issue #2921.
+        error_class = "builder_thinking_stall"
+        details = (
+            "Builder thinking stall retry budget exhausted — extended thinking "
+            "output with zero tool calls detected (fallback cleanup)"
+        )
     else:
         error_class = "builder_unknown_failure"
         # Include post-mortem diagnostics if available (issue #2766).

--- a/loom-tools/tests/shepherd/test_cli.py
+++ b/loom-tools/tests/shepherd/test_cli.py
@@ -4449,6 +4449,60 @@ class TestRecordFallbackFailure:
     @patch("loom_tools.shepherd.cli.get_pr_for_issue", return_value=None)
     @patch("loom_tools.common.systematic_failure.detect_systematic_failure", return_value=None)
     @patch("loom_tools.common.systematic_failure.record_blocked_reason")
+    def test_records_thinking_stall_class(
+        self,
+        mock_record: MagicMock,
+        mock_detect: MagicMock,
+        mock_get_pr: MagicMock,
+    ) -> None:
+        """Should record builder_thinking_stall when abandonment_info has thinking_stall flag (issue #2921)."""
+        ctx = MagicMock()
+        ctx.config.issue = 42
+        ctx.repo_root = Path("/fake/repo")
+        ctx.abandonment_info = {
+            "phase": "builder",
+            "failure_data": {"thinking_stall": True, "exit_code": 1},
+            "message": "builder thinking stall: extended thinking output with zero tool calls detected",
+        }
+
+        _record_fallback_failure(ctx, ShepherdExitCode.BUILDER_FAILED)
+
+        mock_record.assert_called_once_with(
+            Path("/fake/repo"),
+            42,
+            error_class="builder_thinking_stall",
+            phase="builder",
+            details=(
+                "Builder thinking stall retry budget exhausted — extended thinking "
+                "output with zero tool calls detected (fallback cleanup)"
+            ),
+        )
+        mock_detect.assert_called_once_with(Path("/fake/repo"))
+
+    @patch("loom_tools.shepherd.cli.get_pr_for_issue", return_value=None)
+    @patch("loom_tools.common.systematic_failure.detect_systematic_failure", return_value=None)
+    @patch("loom_tools.common.systematic_failure.record_blocked_reason")
+    def test_thinking_stall_without_abandonment_info_falls_through_to_unknown(
+        self,
+        mock_record: MagicMock,
+        mock_detect: MagicMock,
+        mock_get_pr: MagicMock,
+    ) -> None:
+        """Should fall through to builder_unknown_failure when no abandonment_info set."""
+        ctx = MagicMock()
+        ctx.config.issue = 42
+        ctx.repo_root = Path("/fake/repo")
+        ctx.abandonment_info = None
+        ctx.last_postmortem = None
+
+        _record_fallback_failure(ctx, ShepherdExitCode.BUILDER_FAILED)
+
+        call_args = mock_record.call_args
+        assert call_args.kwargs.get("error_class") == "builder_unknown_failure"
+
+    @patch("loom_tools.shepherd.cli.get_pr_for_issue", return_value=None)
+    @patch("loom_tools.common.systematic_failure.detect_systematic_failure", return_value=None)
+    @patch("loom_tools.common.systematic_failure.record_blocked_reason")
     def test_records_unknown_failure_class(
         self,
         mock_record: MagicMock,
@@ -4460,6 +4514,7 @@ class TestRecordFallbackFailure:
         ctx.config.issue = 42
         ctx.repo_root = Path("/fake/repo")
         ctx.last_postmortem = None
+        ctx.abandonment_info = None
 
         _record_fallback_failure(ctx, ShepherdExitCode.BUILDER_FAILED)
 
@@ -4485,6 +4540,7 @@ class TestRecordFallbackFailure:
         ctx = MagicMock()
         ctx.config.issue = 42
         ctx.repo_root = Path("/fake/repo")
+        ctx.abandonment_info = None
         ctx.last_postmortem = {
             "summary": "CLI started but produced zero output; wall: 5s; exit(wait=1)",
         }
@@ -4539,6 +4595,7 @@ class TestRecordFallbackFailure:
         ctx = MagicMock()
         ctx.config.issue = 42
         ctx.repo_root = tmp_path
+        ctx.abandonment_info = None
 
         _record_fallback_failure(ctx, ShepherdExitCode.BUILDER_FAILED)
 
@@ -4569,6 +4626,7 @@ class TestRecordFallbackFailure:
         ctx = MagicMock()
         ctx.config.issue = 42
         ctx.repo_root = tmp_path
+        ctx.abandonment_info = None
         ctx.last_postmortem = None
 
         _record_fallback_failure(ctx, ShepherdExitCode.BUILDER_FAILED)

